crates/exex/exex/src/backfill.rs

```diff
@@ -0,0 +1,306 @@
+use std::{
+    ops::RangeInclusive,
+    time::{Duration, Instant},
+};
+
+/// Factory for creating new backfill jobs.
+#[derive(Debug, Clone)]
+pub struct BackfillJobFactory<Node: FullNodeComponents> {
+    components: Node,
+    config: NodeConfig,
+    thresholds: BackfillThresholds,
+}
+
+impl<Node: FullNodeComponents> BackfillJobFactory<Node> {
+    /// Creates a new backfill job factory.
+    pub const fn new(components: Node, config: NodeConfig, thresholds: BackfillThresholds) -> Self {
+        Self { components, config, thresholds }
+    }
+
+    /// Creates a new backfill job for the given range.
+    pub fn backfill(&self, range: RangeInclusive<BlockNumber>) -> BackfillJob<Node> {
+        BackfillJob {
+            components: self.components.clone(),
+            prune_modes: self.config.prune_config().unwrap_or_default().segments,
+            range,
+            thresholds: self.thresholds.clone(),
+        }
+    }
+}
+
+/// Backfill job started for a specific range.
+///
+/// It implements [`Iterator`] that executes blocks in batches and yields
+/// [notifications](`crate::ExExNotification)s.
+#[derive(Debug)]
+pub struct BackfillJob<Node: FullNodeComponents> {
+    components: Node,
+    prune_modes: PruneModes,
+    range: RangeInclusive<BlockNumber>,
+    thresholds: BackfillThresholds,
+}
+
+impl<Node: FullNodeComponents> Iterator for BackfillJob<Node> {
+    type Item = Result<Chain, BlockExecutionError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.range.is_empty() {
+            return None
+        }
+
+        Some(self.execute_range())
+    }
+}
+
+impl<Node: FullNodeComponents> BackfillJob<Node> {
+    fn execute_range(&mut self) -> Result<Chain, BlockExecutionError> {
+        let provider = self.components.provider();
+        let provider_ro = provider.database_provider_ro()?.disable_long_read_transaction_safety();
+
+        let mut executor = self.components.block_executor().batch_executor(
+            StateProviderDatabase(HistoricalStateProviderRef::new(
+                provider_ro.tx_ref(),
+                *self.range.start(),
+                provider.static_file_provider(),
+            )),
+            self.prune_modes.clone(),
+        );
+
+        let mut fetch_block_duration = Duration::default();
+        let mut execution_duration = Duration::default();
+        let mut cumulative_gas = 0;
+        let batch_start = Instant::now();
+
+        let mut blocks = Vec::new();
+        for block_number in self.range.clone() {
+            // Fetch the block
+            let fetch_block_start = Instant::now();
+
+            let td = provider
+                .header_td_by_number(block_number)?
+                .ok_or_else(|| ProviderError::HeaderNotFound(block_number.into()))?;
+
+            // we need the block's transactions along with their hashes
+            let block = provider
+                .block_with_senders(block_number.into(), TransactionVariant::WithHash)?
```

Comment:
there's even `sealed_block_with_senders_range`

Reply:
added a separate `sealed_block_with_senders`, so we don't create cursors to just query one block